### PR TITLE
sdk-core, node: add file system abstractions

### PR DESCRIPTION
**Pull request by perf2711**
This PR moves file system abstraction to sdk-core, allowing other clients to use this abstraction to achieve, for example, database support, without reimplementing the business logic.

Database file support has also been moved to sdk-core, which works with `FileSystem` interface now.

Jira task: BT-585

### CODE REVIEW

File: packages/node/src/BacktraceClient.ts [hunk 1]
```diff
@@ -219,7 +221,7 @@ export class BacktraceClient extends BacktraceCoreClient {
     }
 
     private async loadNodeCrashes() {
-        if (!this.options.database?.captureNativeCrashes) {
+        if (!this.fileSystem || !this.options.database?.captureNativeCrashes) {
```
Comment (konraddysput) [inline packages/node/src/BacktraceClient.ts:224]: this is out of scope for this review, however I have a question:

- client is a programming interface for developer to use Backtrace. For managing database we have a separated class which is BacktraceDatabase. Shouldn't this code be a part of the database somehow?

File: packages/sdk-core/src/modules/storage/FileSystem.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import { BacktraceAttachment } from '../../model/attachment';
+
+export interface FileSystem {
+    readDir(dir: string): Promise<string[]>;
```
Comment (konraddysput) [inline packages/sdk-core/src/modules/storage/FileSystem.ts:4]: quick idea:

based on the code we had some assumptions or we needed to pass some options - like for example:
- we expect to read the directory and return all files with extensions 
- we also expect to return files in specific string format - utf8

Shouldn't we note those assumptions in the interface to avoid potential surprises later? 
Previous code:

```
  const databaseFiles = fs.readdirSync(databasePath, {
            encoding: 'utf8',
            withFileTypes: true,
        });
```

File: packages/node/src/storage/NodeFileSystem.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import fs from 'fs';
+import { BacktraceFileAttachment } from '../attachment';
+
+export class NodeFileSystem implements FileSystem {
+    public readDir(dir: string): Promise<string[]> {
+        return fs.promises.readdir(dir);
+    }
+
+    public readDirSync(dir: string): string[] {
+        return fs.readdirSync(dir);
```
Comment (konraddysput) [inline packages/node/src/storage/NodeFileSystem.ts:11]: previously you were using the following options to read the dir:

```
  const databaseFiles = fs.readdirSync(databasePath, {
            encoding: 'utf8',
            withFileTypes: true,
        });
```

Are you sure it works correctly without them?

File: packages/sdk-core/src/BacktraceCoreClient.ts [hunk 4]
```diff
@@ -71,11 +77,15 @@ export abstract class BacktraceCoreClient {
     }
 
     public get metrics(): BacktraceMetrics | undefined {
-        return this._metrics;
+        return this._modules.get(BacktraceMetrics);
```
Comment (konraddysput) [inline packages/sdk-core/src/BacktraceCoreClient.ts:80]: did you merge your code from dev to this branch? You already landed this change

File: packages/sdk-core/src/BacktraceCoreClient.ts [hunk 5]
```diff
@@ -131,25 +143,27 @@ export abstract class BacktraceCoreClient {
...
             this._sdkOptions,
             stackTraceConverter,
             this.attributeManager,
             new DebugIdProvider(stackTraceConverter, this._setup.debugIdMapProvider),
         );
 
-        this._attachments = this.options.attachments ?? [];
+        if (this.options?.database?.enable === true) {
+            const provider =
+                this._setup.databaseStorageProvider ??
```
Comment (konraddysput) [inline packages/sdk-core/src/BacktraceCoreClient.ts:156]: You can drop databaseStorageProvider support and use only a new file system. The core client interface is not exposed outside the package and also to construct it you need to have package internals not exposed anywhere else.

File: packages/sdk-core/src/BacktraceCoreClient.ts [hunk 6]
```diff
@@ -158,19 +172,38 @@ export abstract class BacktraceCoreClient {
...
+            this.attributeManager.addProvider(breadcrumbsManager);
+            this._modules.set(BreadcrumbsManager, breadcrumbsManager);
         }
 
-        this.initialize();
         this._enabled = true;
     }
 
+    public initialize() {
+        if (this.fileSystem && this.options.database?.createDatabaseDirectory) {
```
Comment (konraddysput) [inline packages/sdk-core/src/BacktraceCoreClient.ts:189]: I think this should be a part of the BacktraceDatabase implementation.

File: packages/sdk-core/src/modules/database/BacktraceDatabaseFileRecord.ts [hunk 7]
```diff
@@ -24,19 +26,25 @@ export class BacktraceDatabaseFileRecord implements BacktraceDatabaseRecord {
     public static fromRecord(record: BacktraceDatabaseRecord) {
         return new BacktraceDatabaseFileRecord(
             record,
-            record.attachments.filter((n) => n instanceof BacktraceFileAttachment) as BacktraceFileAttachment[],
+            record.attachments.filter(BacktraceDatabaseFileRecord.isFileAttachment),
```
Comment (konraddysput) [inline packages/sdk-core/src/modules/database/BacktraceDatabaseFileRecord.ts:29]: Isn't it cleaner we expect to pass File Attachment here instead of doing this inside the class? Now if someone will use it, it would be surprising for someone attachment are not there or are filtered out by the class implementation. WDYT?

**Review by konraddysput — CHANGES_REQUESTED** [review]
Can we please merge changes from the dev/your previous branch?

File: packages/sdk-core/src/modules/database/BacktraceDatabaseFileStorageProvider.ts [hunk 8]
```diff
@@ -76,48 +65,44 @@ export class BacktraceDatabaseFileStorageProvider implements BacktraceDatabaseSt
...
+            this._fileSystem.unlinkSync(recordPath);
             return true;
         } catch {
             return false;
         }
     }
 
     private getRecordPath(id: string): string {
-        return path.join(this._path, `${id}${this.RECORD_SUFFIX}`);
+        return this._path + '/' + `${id}${this.RECORD_SUFFIX}`;
```
[the remark this answers is not in the document]
Comment (perf2711) [reply packages/sdk-core/src/modules/database/BacktraceDatabaseFileStorageProvider.ts:106]: `path` is unavailable in `sdk-core`

File: packages/sdk-core/src/modules/database/BacktraceDatabaseFileRecord.ts [hunk 9]
```diff
@@ -24,19 +26,25 @@ export class BacktraceDatabaseFileRecord implements BacktraceDatabaseRecord {
...
 
-    public static fromJson(json: string): BacktraceDatabaseFileRecord | undefined {
+    public static fromJson(json: string, fileSystem: FileSystem): BacktraceDatabaseFileRecord | undefined {
         try {
             const record = JSON.parse(json) as BacktraceDatabaseFileRecord;
             const attachments = record.attachments
-                ? record.attachments.map((n) => new BacktraceFileAttachment(n.filePath))
+                ? record.attachments
+                      .filter(BacktraceDatabaseFileRecord.isFileAttachment)
+                      .map((n) => fileSystem.createAttachment(n.filePath, n.name))
```
[the remark this answers is not in the document]
Comment (perf2711) [reply packages/sdk-core/src/modules/database/BacktraceDatabaseFileRecord.ts:39]: we do not have `BacktraceFileAttachment` class here, it is available in Node

File: packages/sdk-core/src/modules/database/BacktraceDatabaseFileRecord.ts [hunk 10]
```diff
@@ -24,19 +26,25 @@ export class BacktraceDatabaseFileRecord implements BacktraceDatabaseRecord {
...
                 : [];
             return new BacktraceDatabaseFileRecord(record, attachments);
         } catch {
             return undefined;
         }
     }
+
+    private static isFileAttachment(attachment: BacktraceAttachment): attachment is BacktraceFileAttachment {
```
[the remark this answers is not in the document]
Comment (perf2711) [reply packages/sdk-core/src/modules/database/BacktraceDatabaseFileRecord.ts:47]: I can pass it more safely to `filter` as a function and not worry about it accessing `this`.

File: packages/sdk-core/src/BacktraceCoreClient.ts (hunk 4 above)
Comment (perf2711) [reply packages/sdk-core/src/BacktraceCoreClient.ts]: due to merging previous stuff, I had to rebase.

File: packages/sdk-core/src/modules/storage/FileSystem.ts (hunk 2 above)
Comment (perf2711) [reply packages/sdk-core/src/modules/storage/FileSystem.ts:4]: So `withFileTypes` returns not extensions, but objects instead that have `isFile` or `isDir` methods. I opted for skipping that as we will have no directories being created with the names that we search for.

Also, `utf8` is default:
image

File: packages/node/src/storage/NodeFileSystem.ts (hunk 3 above)
Comment (perf2711) [reply packages/node/src/storage/NodeFileSystem.ts:11]: yes

File: packages/sdk-core/src/BacktraceCoreClient.ts (hunk 5 above)
Comment (perf2711) [reply packages/sdk-core/src/BacktraceCoreClient.ts:156]: I'm not sure what you mean? Use `fileSystem` directly in `BacktraceDatabase`?

File: packages/sdk-core/src/modules/database/BacktraceDatabaseFileRecord.ts (hunk 7 above)
Comment (perf2711) [reply packages/sdk-core/src/modules/database/BacktraceDatabaseFileRecord.ts:29]: not sure what you mean by this. I don't have `BacktraceFileAttachment` implementation here. But every implementatio of the interface will return `true` from this filter.

File: packages/node/src/BacktraceClient.ts [hunk 11]
```diff
@@ -36,7 +36,7 @@ export class BacktraceClient extends BacktraceCoreClient {
             breadcrumbsSetup: {
                 subscribers: breadcrumbsEventSubscribers,
             },
-            databaseStorageProvider: BacktraceDatabaseFileStorageProvider.createIfValid(options.database),
+            fileSystem: new NodeFileSystem(),
```
[the remark this answers is not in the document]
Comment (perf2711) [reply packages/node/src/BacktraceClient.ts]: Done

**Comment by perf2711** [comment]
> Can we please merge changes from the dev/your previous branch?

Merged.

File: packages/node/src/BacktraceClient.ts (hunk 1 above)
Comment (perf2711) [reply packages/node/src/BacktraceClient.ts:225]: I think not, because depending all stuff on BacktraceDatabase that can be dependent just on FileSystem is wrong. It's completely unnecessary.

It makes sense for me that FileSystem is the dependency of BacktraceDatabase and other stuff that may use it, and not have the DB inbetween.
image

File: packages/sdk-core/src/BacktraceCoreClient.ts (hunk 6 above)
Comment (perf2711) [reply packages/sdk-core/src/BacktraceCoreClient.ts:189]: This is really troublesome. If we didn't rely on a database directory for everything, but just use something like "storageOptions", this would make perfect sense.

Moving it to database will cause undeterministic behavior, when some module depending on DB's directory will be initialized before DB.

File: packages/sdk-core/src/BacktraceCoreClient.ts (hunk 5 above)
Comment (konraddysput) [reply packages/sdk-core/src/BacktraceCoreClient.ts:156]: I think you're trying to use databaseStorageProvider - I think you can remove it and use only the new file system variable

File: packages/node/src/BacktraceClient.ts (hunk 1 above)
Comment (konraddysput) [reply packages/node/src/BacktraceClient.ts:225]: but here you have something that clearly depends on the database. You literally need to check if the database is initialized and do something only when it is.

File: packages/node/src/storage/NodeFileSystem.ts (hunk 3 above)
Comment (konraddysput) [reply packages/node/src/storage/NodeFileSystem.ts:11]: How so - we had fun use cases with our customers in India/asia having issues with default encoding. I think using utf8 is necessary. is this a default option?

File: packages/node/src/BacktraceClient.ts (hunk 1 above)
Comment (perf2711) [reply packages/node/src/BacktraceClient.ts:225]: that's a business decision, not a code decision.

File: packages/node/src/storage/NodeFileSystem.ts (hunk 3 above)
Comment (perf2711) [reply packages/node/src/storage/NodeFileSystem.ts:11]: `utf-8` is the default encoding in `readdir`/`readdirSync`

File: packages/sdk-core/src/BacktraceCoreClient.ts (hunk 5 above)
Comment (perf2711) [reply packages/sdk-core/src/BacktraceCoreClient.ts:156]: so we will not be planning to override providers from outside, that's what you mean?

Comment (konraddysput) [reply packages/sdk-core/src/BacktraceCoreClient.ts:156]: is there any reason to do that if we have file system? I think, no

Comment (perf2711) [reply packages/sdk-core/src/BacktraceCoreClient.ts]: Ok, removed

**Review by konraddysput — APPROVED** [review]
